Fix memory and other explanation (#46)
owncloud/docs-ocis: modules/ROOT/pages/prerequisites/index.adoc

{description}

> Note that oCIS is based on a highly dynamic interoperable _microservices_ concept, even when it is delivered as single binary. Of course this software has to run on hardware, but its ability to adapt to various and maybe dynamic load scenarios across physical boundaries means that you can setup your physical environment flexibly using modern technology.

Reviewer: ```suggestion
Note that oCIS is based on a highly dynamic interoperable _microservices_ concept, although it is delivered as a single binary. Of course, this software has to run on hardware, but its ability to adapt to various and maybe dynamic load scenarios across physical boundaries means that you can set up your physical environment flexibly, using modern technology.
```

> Note that oCIS is based on a highly dynamic interoperable _microservices_ concept, even it is delievered as single binary. Of course this software has to run on hardware, but its ability to adopt to various and maybe dynamic load scenarious across physical bounderies means, that you can setup your physical environment accordingly using modern technology.

Reviewer 2: ```suggestion
Note that oCIS is based on a highly dynamic interoperable _microservices_ concept, even when it is delivered as single binary. Of course this software has to run on hardware, but its ability to adapt to various and maybe dynamic load scenarios across physical boundaries means that you can setup your physical environment flexibly using modern technology.
```

> The following sections can be seen as a fundament to understand general prerequisites, but their implementation is very dependent on your usecase. To support your decision about the environment, we recommend to read the documents about  xref:architecture/index.adoc[Architecture and Concepts] and xref:availability/index.adoc[Availability, Redundancy and Load Distribution] first.

Reviewer 2: ```suggestion
The following sections are fundamental to understanding the general prerequisites, but their implementation is very dependent on your use-case. To support your decision about the environment, we recommend reading the documents about  xref:architecture/index.adoc[Architecture and Concepts] and xref:availability/index.adoc[Availability, Redundancy and Load Distribution] first.
```

== Hardware

> The minimum hardware requirements depend on the usage scenario you have in mind for your oCIS instance. You need at least hardware that runs a Linux based operating system, preferrable 64bit.

Reviewer 2: ```suggestion
The minimum hardware requirements depend on the usage scenario you have in mind for your oCIS instance. You need at least hardware that runs a Linux based operating system, preferably 64bit.
```

Hmmm - is it possible to run oCIS on some 32-bit hardware?

* For simple testing purposes, you can even use a laptop or Raspberry Pi 4 (ARM64) with a minimum of 512MB of RAM.
* For more intense testing purposes, the recommended start hardware is a multi-core processor with at least 4GB of RAM.

> * For production environments, CPU and RAM are the main limiting factors. A clear guideline for CPU's cant be given at the moment as there are too many influencing factors. See xref:ram-considerations[RAM Considerations] for memory and xref:bandwidth-considerations[Bandwidth Considerations] for network related factors.

Reviewer: ```suggestion
* For production environments, CPU and RAM are the main limiting factors. A clear guideline for CPUs can't be provided at this time since there are too many influencing factors. See xref:ram-considerations[RAM Considerations] for memory and xref:bandwidth-considerations[Bandwidth Considerations] for network-related factors.
```

> * For productive environments, CPU and RAM are the main limiting factors. A clear guideline for CPU's cant be given at the moment as there are too many influencing factors. See xref:ram-considerations[RAM Considerations] for memory and xref:bandwidth-considerations[Bandwidth Considerations] for network related factors.

Reviewer 2: ```suggestion
* For production environments, CPU and RAM are the main limiting factors. A clear guideline for CPU's cant be given at the moment as there are too many influencing factors. See xref:ram-considerations[RAM Considerations] for memory and xref:bandwidth-considerations[Bandwidth Considerations] for network related factors.
```

=== RAM Considerations

// harvested from https://owncloud.dev/architecture/efficient-stat-polling/
// text adopted based on an intense discussion with jfd/willy on 11.3.

// fixme: how to read stat info (local and eg nfs)

> RAM requirements are mainly based on caching the mounted filesystems meta-data like inodes (even extended attributes are supported) and include calculated ETags (entity tag). Meta-data caching is mostly based on the kernal caching capability of the OS for the mounted filesystems. Dependent on the filesystem(s) mounted, a different block size for the meta-data may be used. Avoid swap space usage because of meta-data caching for performance reasons.

Reviewer 2: ```suggestion
RAM requirements are mainly based on caching the mounted filesystems meta-data like inodes (even extended attributes are supported) and include calculated ETags (entity tag). Meta-data caching is mostly based on the kernel caching capability of the OS for the mounted filesystems. Dependent on the filesystem(s) mounted, a different block size for the meta-data may be used. Avoid swap space usage because of meta-data caching for performance reasons.
```

> NOTE: In an optimal environment, all the meta-data is kept in memory for best perfomance. If that information is not available in memory, physical access to the storage backend needs to be made to gather and calculate the required information slowing down the response. In real life, a balance between memory for cached meta-data, physical storage access and performance will take place.

Reviewer: ```suggestion
NOTE: In an ideal environment, all the meta-data is kept in memory for best performance. If that information is not available in memory, physical access to the storage backend is necessary to gather and calculate the required information, which slows down the response. In real life, memory for cached meta-data, physical storage access and performance will have to be balanced.
```

NOTE: Your RAM requirements may change over time depending on the amount of users accessing an increasing amount of data from different sources with different change rates concurrently.

==== Calculation example:

For a single machine using a local posix storage, the Linux kernel caches meta-data which is necessary to calculate the ETag. Note that the following description focuses on classical meta-data. Additional information added to the meta-data by oCIS like user added tags or comments are not considered. This functionality may increase the meta-data quantity that should be cached.

> // fixme: differnt fs have different fs have different meta-data blocksizes or can even chain them

Reviewer 2: ```suggestion
// fixme: different fs have different meta-data block-sizes or can even chain them
```

> With a meta-data size of 4K each, as it it used in the EXT4 filesystem, the following RAM is needed to completely cache all meta-data:

Reviewer 2: ```suggestion
With a meta-data size of 4K each, as is used in the EXT4 filesystem, the following RAM is needed to completely cache all meta-data:
```

[width="60%",cols="^40%,^70%",options="header"]
|===
| Meta-data items
a| RAM requirement for [.underline]#_caching only_#

| 256
| 1 MB

| 1 K
| 4 MB

| 1 M
| 4 GB
|===

> Using the above table and some sample environments from below, you can easily match the sweet spot and possibly extrapolate it to your environment. Consider that more users will do more requests for content or changes regularly, independent of if they are accessing oCIS via the web interface, via a mobile client, or the desktop synchronisation client. If this access is not held in memory, the access needs to go thru the backend slowing down the overall response and consumes more CPU accordingly.

Reviewer: ```suggestion
Using the above table and some sample environments from below, you can easily match the sweet spot and possibly extrapolate it to your environment. Consider that more users will do more requests for content or changes regularly, independent of whether they are accessing oCIS via the web interface, via a mobile client or the desktop synchronization client. If this access is not held in memory, the access needs to go through the backend slowing down the overall response and consumes more CPU accordingly.
```

> Using the above table and some sample environments from below, you can easily match the sweet spot and possibly extrapolate it to your environment. Consider that more users will do more requests for content or changes regulary, independent if they are accessing oCIS via the web interface, via a mobile client, or the desktop synchronisation client. If this access is not held in memory, the access needs to go thru the backend slowing down the overall response and consumes more CPU accordingly.

Reviewer 2: ```suggestion
Using the above table and some sample environments from below, you can easily match the sweet spot and possibly extrapolate it to your environment. Consider that more users will do more requests for content or changes regularly, independent of if they are accessing oCIS via the web interface, via a mobile client, or the desktop synchronisation client. If this access is not held in memory, the access needs to go thru the backend slowing down the overall response and consumes more CPU accordingly.
```

> Note that the table below shows only one xref:background[spaces] created and no shares or reshares which would multiply the needed quantity of meta-data.

Reviewer: ```suggestion
Note that the table below shows only one xref:background[space] created and no shares or re-shares which would multiply the needed quantity of meta-data.
```

[[sample_environments]]
[caption=]
.Sample environments (Node = file or folder)
[width="100%",cols="70%,^40%,70%,^40%,70%,^40%",options="header"]
|===
2+^| Environment 1
2+^| Environment 2
2+^| Environment 3

a| *Nodes*
| 3 Millions
a| *Nodes*
| 2 Millions
a| *Nodes*
| 3 Billions

a| *Users*
| 30
a| *Users*
| 80
a| *Users*
| 500

a| *Changed Nodes/h*
| 65
a| *Changed Nodes/h*
| 18
a| *Changed Nodes/h*
| 178

a| *Changed meta/h*
| 260
a| *Changed meta/h*
| 72
a| *Changed meta/h*
| 712
|===

> It is assumed in the examples above, that on avarage, the original change was 4 levels below the root level of a single spaces, creating 4 times the quantity of changed meta-data.

Reviewer 2: ```suggestion
It is assumed in the examples above, that on avarage, the original change was 4 levels below the root level of a single space, creating 4 times the quantity of changed meta-data.
```

When there is one instance of "spaces", is it a "single space"?

==== Background

> Storing meta-data in memory is important with respect to access and synchronisation performance. oCIS has the concept of _spaces_ which you can see in a nutshell as a space containing files and folders which is not owned by a user by default, but gets assigned to users. Spaces can be the users home, any mount created or an unlimited quantity of shares and reshares.

Reviewer: ```suggestion
Storing meta-data in memory is important with respect to access and synchronization performance. oCIS has the concept of _spaces_ which you can see in a nutshell as a space containing files and folders which is not owned by a user by default, but gets assigned to users. Spaces can be the users home, any mount created or an unlimited quantity of shares and re-shares.
```

> The reason storing meta-data in memory is important with respect to access and synchronisation performance. oCIS has the concept of _spaces_ which you can see in a nutshell as a space containing files and folders which is not owned by a user by default, but gets assigned to users. Spaces can be the users home, any mount created or an unlimited quantity of shares and reshares.

Reviewer 2: ```suggestion
Storing meta-data in memory is important with respect to access and synchronisation performance. oCIS has the concept of _spaces_ which you can see in a nutshell as a space containing files and folders which is not owned by a user by default, but gets assigned to users. Spaces can be the users home, any mount created or an unlimited quantity of shares and reshares.
```

> Each space provides a virtual root and occupies 4K for this root meta-data. Additionally each space provides the complete meta-data information about the underlaying content. As you can see, the quantity of meta-data can grow high.

Reviewer 2: ```suggestion
Each space provides a virtual root and occupies 4K for this root meta-data. Additionally each space provides the complete meta-data information about the underlying content. As you can see, the quantity of meta-data can grow large.
```

Whenever a file or folder (node) changes its content or metadata somewhere in the path, the ETag gets recalculated and the change is propagated to the spaces root for identifying a change. To let a client detect changes, the discovery phase starts at the spaces root and checks if the ETag has changed since the last discovery going forward in the tree of changed ETags.

Backend Check::
Checks for changes in the backend are made every 30 seconds for the available spaces. As the space is owner-less, the check appears only once in that interval. The value is configurable to adapt to storage IO load.

Client Check::

> Usually, every connected client polls his assigned spaces root ETag also every 30 seconds and compares to the former ETag recieved for changes. Based on detected changes, actions take place.

Reviewer: ```suggestion
Usually, every connected client also polls his assigned spaces root ETag every 30 seconds and compares it to the former ETag received for changes. Based on detected changes, actions take place.
```

> This makes it clear why RAM is an essential performance factor for client access and synchronisation when more spaces are present. If the gathered spaces root ETag has changed and the underlying meta-data is not in memory, it has to be queried from the physical storage and the ETag needs recalculation. 

Reviewer: ```suggestion
This makes it clear why RAM is an essential performance factor for client access and synchronization when more spaces are present. If the gathered spaces root ETag has changed and the underlying meta-data is not in memory, it has to be queried from the physical storage and the ETag needs recalculation. 
```

> This makes it clear why RAM is an essential performance factor for client access and synchronisation when more spaces are present. If the gathered spaces root ETag has changed and the underlaying meta-data is not in memory, it has to be queried from the physical storage and the ETag needs recalculation. 

Reviewer 2: ```suggestion
This makes it clear why RAM is an essential performance factor for client access and synchronisation when more spaces are present. If the gathered spaces root ETag has changed and the underlying meta-data is not in memory, it has to be queried from the physical storage and the ETag needs recalculation. 
```

// fixme: see https://rudd-o.com/linux-and-free-software/tales-from-responsivenessland-why-linux-feels-slow-and-how-to-fix-that
// fixme: https://manhart.blog/2020/04/linux-leistungsverbesserungen/
// fixme: personal note: reducing vfs_cache_pressure from 100 to 20 improved the "real" component by 40% up to 90%!

=== Bandwidth Considerations

The bandwidth requirements and limitations are based on the following background. Note that this is a view on the internal network (LAN) only. Any considerations about access from the Internet are not part of this description but can be derived from the LAN point of view:

Clients, which are accessing oCIS, request information about what has changed. Depending on the response and if a file synchronization is required, different bandwidth needs may result. Note that when using e.g. the Desktop Client and virtual files (VFS), only those files get physically synced which are set to be locally present, preventing additional bandwidth consumption.

// fixme: the bandwidth calculation in the devdocs is imho wrong as the bigger number is the response and not the request which is then the delimiting factor

Request for changed elements::

> To get the information about changes, the request always starts at the spaces root looking for changed ETags and goes along only that path that has changed elements. Therefore PROPFIND requests and responses are used. A request has about 500 bytes and a response has roughly 800 bytes in size.

Reviewer: ```suggestion
To get the information about changes, the request always starts at the spaces root, looking for changed ETags, and follows only a path that has changed elements. Therefore PROPFIND requests and responses are used. A request has about 500 bytes and a response has roughly 800 bytes in size.
```

Request syncing changed files::
When a file has been identified to be physically synced, the bandwidth requirements depend on the size and the time it should finish. Note that syncing changed files can saturate a network more easily than the handling of changed ETags!

Calculation example::
// fixme: check calculation
+

> Consider the xref:sample_environments[sample environments table] above to see how clients will trigger propfind responses resulting in bandwidth consumption. 500 concurring syncing users, syncing with the default setting of every 30 sec, will create about ~3K propfind requests (500 x 712 / 60 / 2) which consume about 2.4MB/s of bandwidth (3K x 800B) - without doing the file syncs necessary. The physical transfer will create extra bandwidth requirements.

Reviewer: ```suggestion
Consider the xref:sample_environments[sample environments table] above to see how clients will trigger PROPFIND responses resulting in bandwidth consumption. 500 concurring syncing users, syncing with the default setting of every 30 sec, will create about ~3K PROPFIND requests (500 x 712 / 60 / 2) which consume about 2.4MB/s of bandwidth (3K x 800B) - without doing the file syncs necessary. The physical transfer will create extra bandwidth requirements.
```

> Consider the xref:sample_environments[sample environments table] above to see how clients will trigger propfind responses resulting in bandwidth consumption. 500 concurring syncing users, syncinging with the default setting of every 30 sec, will create about ~3K propfind requests (500 x 712 / 60 / 2) which consume about 2.4MB/s of bandwidth (3K x 800B) - without doing the file syncs necessary. The physical transfer will create extra bandwidth requirements.

Reviewer 2: ```suggestion
Consider the xref:sample_environments[sample environments table] above to see how clients will trigger propfind responses resulting in bandwidth consumption. 500 concurring syncing users, syncing with the default setting of every 30 sec, will create about ~3K propfind requests (500 x 712 / 60 / 2) which consume about 2.4MB/s of bandwidth (3K x 800B) - without doing the file syncs necessary. The physical transfer will create extra bandwidth requirements.
```
